### PR TITLE
Add google test recipe

### PR DESCRIPTION
**Pull request by SylvainCorlay**
Closes https://github.com/conda-forge/staged-recipes/pull/623

### CODE REVIEW

**Comment by jakirkham** [comment]
So there is also this recipe for google test ( https://github.com/conda-forge/staged-recipes/pull/623 ). Maybe we can combine the efforts somehow.

cc @scopatz

**Comment by SylvainCorlay** [comment]
> So there is also this recipe for google test ( #623 ). Maybe we can combine the efforts somehow.
> cc @scopatz

Sorry I did not see this one. :+1: on joining forces on either PR.

**Comment by SylvainCorlay** [comment]
@scopatz would like to be added as a maintainer to this recipe?

**Comment by scopatz** [comment]
Sure you can add me if you'd like.

File: recipes/gtest/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+source:
+  url: https://github.com/google/googletest/archive/release-{{ version }}.tar.gz
+  fn: release-{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
+
+requirements:
+  build:
```
Comment (jakirkham) [inline recipes/gtest/meta.yaml]: Please add `toolchain`.

File: recipes/gtest/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+{% set version = "1.7.0" %}
+{% set sha256 = "f73a6546fdf9fce9ff93a5015e0333a8af3062a152a9ad6bcb772c96687016cc" %}
+
+package:
+  name: gtest
+  version: {{ version }}
+
+source:
+  url: https://github.com/google/googletest/archive/release-{{ version }}.tar.gz
+  fn: release-{{ version }}.tar.gz
```
Comment (jakirkham) [inline recipes/gtest/meta.yaml]: Please move `fn` before `url`.

File: recipes/gtest/build.sh [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+cmake $SRC_DIR/CMakeLists.txt
+make
```
Comment (jakirkham) [inline recipes/gtest/build.sh]: Is there some kind of `make check` we can run?

File: recipes/gtest/bld.bat [hunk 4]
```diff
@@ -0,0 +1,6 @@
+cmake -G "NMake Makefiles" -D CMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% -D CMAKE_BUILD_TYPE=Release .\CMakeLists.txt
+
+nmake
```
Comment (jakirkham) [inline recipes/gtest/bld.bat]: Is there some kind of `ctest` we can run?

File: recipes/gtest/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    - python
+    - cmake
+
+test:
+  commands:
+    - test -f ${PREFIX}/lib/libgtest.a  # [unix]
+
+about:
+  home: https://github.com/google/googletest
+  license: Apache License, Version 2.0
```
Comment (jakirkham) [inline recipes/gtest/meta.yaml]: Please shorten to `Apache 2.0`.

File: recipes/gtest/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  number: 0
+
+requirements:
+  build:
+    - python
+    - cmake
+
+test:
+  commands:
+    - test -f ${PREFIX}/lib/libgtest.a  # [unix]
```
Comment (jakirkham) [inline recipes/gtest/meta.yaml]: Test for `libgtest_main.a`? Test for `include/gtest`? Windows tests?

File: recipes/gtest/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,33 @@
+{% set version = "1.7.0" %}
```
Comment (jakirkham) [inline recipes/gtest/meta.yaml]: `1.8.0` is out. Though it combines `gtest` and `gmock`.

File: recipes/gtest/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+{% set sha256 = "f73a6546fdf9fce9ff93a5015e0333a8af3062a152a9ad6bcb772c96687016cc" %}
+
+package:
+  name: gtest
```
Comment (jakirkham) [inline recipes/gtest/meta.yaml]: Should it be `googletest` or `gtest`?

File: recipes/gtest/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    - cmake
+
+test:
+  commands:
+    - test -f ${PREFIX}/lib/libgtest.a  # [unix]
+
+about:
+  home: https://github.com/google/googletest
+  license: Apache License, Version 2.0 
+  license_family: Apache
```
Comment (jakirkham) [inline recipes/gtest/meta.yaml]: Please add the license file (e.g. `license_file: LICENSE`).

File: recipes/gtest/meta.yaml (hunk 5 above)
Comment (jakirkham) [inline recipes/gtest/meta.yaml]: Actually this should be [BSD 3-Clause]( https://github.com/google/googletest/blob/release-1.7.0/LICENSE ).

File: recipes/gtest/meta.yaml (hunk 9 above)
Comment (jakirkham) [inline recipes/gtest/meta.yaml]: Also `license_family: BSD`.

File: recipes/gtest/build.sh [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+cp libgtest.a $PREFIX/lib/
+cp libgtest_main.a $PREFIX/lib/
```
Comment (jakirkham) [inline recipes/gtest/build.sh]: What about dynamic libraries?

File: recipes/gtest/meta.yaml (hunk 7 above)
Comment (SylvainCorlay) [inline recipes/gtest/meta.yaml]: Yes, I am not sure that it is the stable version.

File: recipes/gtest/meta.yaml (hunk 8 above)
Comment (SylvainCorlay) [inline recipes/gtest/meta.yaml]: I opened an issue about that :) the debian package name is `libgtest`.

File: recipes/gtest/build.sh (hunk 10 above)
Comment (SylvainCorlay) [inline recipes/gtest/build.sh]: I don't think that the default gtest make file produces any.

File: recipes/gtest/meta.yaml (hunk 6 above)
Comment (SylvainCorlay) [inline recipes/gtest/meta.yaml]: Added a test for `libgtest_main.a` and `include/gtest` directory. No windows test for now.

**Comment by SylvainCorlay** [comment]
@scopatz @jakirkham do you have more comments on the state of the recipe?
Are we good to get this one in?

File: recipes/gtest/meta.yaml (hunk 1 above)
Comment (SylvainCorlay) [inline recipes/gtest/meta.yaml]: done

File: recipes/gtest/meta.yaml (hunk 6 above)
Comment (SylvainCorlay) [inline recipes/gtest/meta.yaml]: done and done

File: recipes/gtest/build.sh (hunk 10 above)
Comment (scopatz) [inline recipes/gtest/build.sh]: There is a way to get it to, and I really think that this should.  Please see the other PR.

**Comment by scopatz** [comment]
The dynamic libraries issues was the reason that I put in the original PR, so I think that should be included here as the package is incomplete for my original use case without it.

**Comment by scopatz** [comment]
Thanks for pushing this forward @SylvainCorlay!

**Comment by SylvainCorlay** [comment]
> The dynamic libraries issues was the reason that I put in the original PR, so I think that should be included here as the package is incomplete for my original use case without it.

Ok, looking into it.

**Comment by SylvainCorlay** [comment]
So the windows build is in fact not successful in the case of the dynamic library.

**Comment by scopatz** [comment]
I am not worried about Windows, personally, so if the dynamic libraries are only on POSIX, that is fine with me.

**Comment by SylvainCorlay** [comment]
OK, I will remove the build of the dynamic libraries on windows for now.

**Comment by SylvainCorlay** [comment]
All set. I can try to enable the windows build in a feedstock PR.

**Comment by jakirkham** [comment]
SGTM 😄

**Comment by jakirkham** [comment]
As for the Google Test recipe name, I did a quick google of the main package managers. CentOS, Arch, Gentoo, and SUSE all do `gtest`. MacPorts does `gtest` and `google-test`, which they have been trying to fix and/or remove both. Homebrew doesn't package it because of [Google's recommendation not to]( https://github.com/google/googletest/blob/d225acc90bc3a8c420a9bcd1f033033c1ccd7fe0/googletest/docs/V1_7_FAQ.md#why-is-it-not-recommended-to-install-a-pre-compiled-copy-of-google-test-for-example-into-usrlocal ). We already decided to package it after some discussion awhile back. Given most people go with `gtest` and that matches with what we normally do, I'm ok with sticking with that if everyone else is.

File: recipes/gtest/meta.yaml (hunk 7 above)
Comment (jakirkham) [inline recipes/gtest/meta.yaml]: Totally ok with starting with this version.

File: recipes/gtest/build.sh [hunk 11]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+# Copy headers
+cp -r $SRC_DIR/include/gtest $PREFIX/include/
+
+# Build and copy static libraries
+mkdir build_static
+cd build_static
+cmake $SRC_DIR
+make
```
Comment (jakirkham) [inline recipes/gtest/build.sh]: Sorry have forgotten. Did we try `make check` or similar?

File: recipes/gtest/build.sh [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  LIBEXT="dylib"
+else
+  # for Linux
+  LIBEXT="so"
+fi
+
+mkdir build_dynamic
+cd build_dynamic
+cmake $SRC_DIR -Dgtest_build_tests=ON
+make
```
Comment (jakirkham) [inline recipes/gtest/build.sh]: Sorry have forgotten. Did we try `make check` or similar?

**Comment by jakirkham** [comment]
Couple questions about running their test suite. Otherwise LGTM and am happy to merge.

**Comment by SylvainCorlay** [comment]
:shipit:

File: recipes/gtest/build.sh (hunk 12 above)
Comment (SylvainCorlay) [inline recipes/gtest/build.sh]: I missed that. Trying now.

Comment (SylvainCorlay) [inline recipes/gtest/build.sh]: There is no check target.

Comment (jakirkham) [inline recipes/gtest/build.sh]: That's right. I had forgotten. Thanks for looking into it.

**Comment by jakirkham** [comment]
Thanks @SylvainCorlay and @scopatz.

File: recipes/gtest/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  name: gtest
+  version: {{ version }}
+
+source:
+  fn: release-{{ version }}.tar.gz
+  url: https://github.com/google/googletest/archive/release-{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
```
Comment (jakirkham) [inline recipes/gtest/meta.yaml]: Oops, we forgot the VC features. 😞 Will add a PR to fix this.

Comment (jakirkham) [inline recipes/gtest/meta.yaml]: Please see PR ( https://github.com/conda-forge/gtest-feedstock/pull/1 ).

**Pull request merged**